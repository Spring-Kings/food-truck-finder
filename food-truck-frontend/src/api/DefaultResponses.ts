import Router from "next/router";

export const DEFAULT_OK_RESP = (_data: any) => {};
export const DEFAULT_ERR_RESP = (err: any) => console.log(err);
export const DEFAULT_ERR_KICK = (err: any) => {
  // Temporary measure: kick back to home
  console.log(err);
  Router.replace("/");
};
