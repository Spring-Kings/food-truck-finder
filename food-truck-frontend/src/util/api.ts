import axios from 'axios';

const api = axios.create({
    baseURL: process.env.FOOD_TRUCK_API_URL,

});

api.interceptors.request.use(
    config => {
        const token = localStorage.getItem('authToken');
        if (token) {
            config.headers["Authorization"] = token;
        }
        const lat = localStorage.getItem("latitude");
        const lng = localStorage.getItem("longitude");
        if (lat && lng) {
            config.headers["Coordinates"] = lat + "," + lng;
        }
        return config;
    }
);


export default api;