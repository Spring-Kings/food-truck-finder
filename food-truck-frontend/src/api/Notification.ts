import api from "../util/api";
import {TruckState} from "../components/TruckView";

export interface Notification {
  id: number;
  message: string;
  truck: TruckState;
  time: Date;
  read: boolean;
}

export const getNotifications = async () => {
  let notifications: Notification[] = [];

  await api.get(`/notifications`, {})
    .then((response: any) => {
      if (response.data) {
        notifications = response.data;
      }
    });

  return notifications;
}

export const sendNotification = async (
  truckId: number,
  message: string,
  onFail?: (res: any) => void
) => {
  let config = {
    truckId: truckId,
    message: message
  };
  await api.post(`/truck/notification`, config)
    .catch(onFail);
}

export const deleteNotification = async (
  notificationId: number,
  onFail?: (res: any) => void
) => {
  await api.delete(`/notification/${notificationId}/delete`, {})
    .catch(onFail);
}

export const setNotificationAsRead = async (
  notificationId: number,
  isRead: boolean,
  onFail?: (res: any) => void
) => {
  await api.put(`/notification/read`, {
    notificationId: notificationId,
    isRead: isRead
  })
    .catch(onFail);
}