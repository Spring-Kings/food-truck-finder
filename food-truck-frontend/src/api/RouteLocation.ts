import api from "../util/api";
import {
  backendToFrontend,
  frontendToBackend,
  RouteLocation,
} from "../components/map/route-map/RouteLocation";

export const loadTodaysRoute = async (
  truckId: number,
  onFail?: (res: any) => void
) => {
  var routePts: RouteLocation[] = [];
  var nextStopId: number = 1;

  await api
    .request({
      url: `/truck/${truckId}/active-route`,
      method: "GET",
    })
    .then((response) => {
      if (response.data)
        routePts = response.data.map((pt: any) =>
          backendToFrontend(pt, nextStopId++)
        );
    })
    .catch(onFail);
  return routePts;
};

export const loadRouteLocations = async (
  routeId: number,
  onFail?: (res: any) => void
) => {
  var routePts: RouteLocation[] = [];
  await api
    .request({
      url: `/truck/route/locations/${routeId}`,
      method: "GET",
    })
    .then((response: any) => {
      if (response.data != undefined) {
        var nextStopId: number = 1;
        routePts = response.data.map((pt: any) =>
          backendToFrontend(pt, nextStopId++)
        );
      }
    })
    .catch(onFail);

  return routePts;
};

export const updateRouteLocations = async (
  routeId: number,
  routePts: RouteLocation[],
  onSuccess?: (res: any) => void,
  onFail?: (res: any) => void
) => {
  await api
    .request({
      url: `/truck/route/locations/${routeId}`,
      data: routePts.flatMap((pt) => frontendToBackend(pt, routeId)),
      method: "POST",
    })
    .then(onSuccess)
    .catch(onFail);
};

export const deleteRouteLocations = async (
  routeId: number,
  trashedPts: RouteLocation[],
  onSuccess?: (res: any) => void,
  onFail?: (res: any) => void
) => {
  await api
    .request({
      url: `/truck/route/locations/${routeId}`,
      data: trashedPts.flatMap((pt) => pt.routeLocationId),
      method: "DELETE",
    })
    .then(onSuccess)
    .catch(onFail);
};
