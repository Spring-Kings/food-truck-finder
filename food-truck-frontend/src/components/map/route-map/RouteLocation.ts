import {LatLngLiteral} from "@google/maps";
import {utcTimeStringToDate} from "../../../util/date-conversions";

/**
 * CREATED: Just created, not sent to backend
 * PERSISTED: Loaded from backend
 * DELETED: PERSISTED and deleted in the frontend
 */
export enum RouteLocationState {
  CREATED,
  PERSISTED,
  DELETED,
}

// Diff from backend structure:
// - Missing routeId
// / Lat/Lng => LatLngLiteral
// + Added stopId
// + Added state
export interface RouteLocation {
  routeLocationId: number;
  arrivalTime: Date;
  exitTime: Date;

  stopId: number;
  coords: LatLngLiteral;
  state: RouteLocationState;

  readonly [x: string]: number | LatLngLiteral | Date | RouteLocationState;
}

export const blankRouteLocation = () => ({
  routeLocationId: -1,
  arrivalTime: new Date(),
  exitTime: new Date(),
  stopId: 0,
  coords: { lat: 0, lng: 0},
  state: RouteLocationState.CREATED,
})

export const backendToFrontend = (pt: any, stopId: number) => {
  // Addition to enable testing--otherwise, milliseconds unpredictably set
  let arrival = utcTimeStringToDate(pt.arrivalTime)
  let exit = utcTimeStringToDate(pt.exitTime)
  arrival?.setMilliseconds(0)
  exit?.setMilliseconds(0)

  // Return mapping
  return {
    stopId: stopId,
    routeLocationId: pt.routeLocationId,
    coords: {
      lat: pt.position.latitude,
      lng: pt.position.longitude,
    },
    arrivalTime: arrival,
    exitTime: exit,
    state: RouteLocationState.PERSISTED,
  }
}

export const frontendToBackend = (pt: RouteLocation, routeId: number) => {
  let result: any = {
    routeId: routeId,
    routeLocationId: pt.routeLocationId > 0 ? pt.routeLocationId : null,
    arrivalTime: pt.arrivalTime,
    exitTime: pt.exitTime,
    lat: pt.coords.lat,
    lng: pt.coords.lng,
  };
  return result;
}
