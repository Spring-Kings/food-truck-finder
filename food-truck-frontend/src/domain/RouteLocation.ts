import {utcTimeStringToDate} from "../util/date-conversions";
import * as t from 'io-ts'
import DateMeta from "../metaclasses/DateMeta";
import LatLngLiteralMeta from "../metaclasses/LatLngLiteralMeta";

export const RouteLocationMeta = t.type({
  routeLocationId: t.number,
  arrivalTime: DateMeta,
  exitTime: DateMeta,
  stopId: t.number,
  coords: LatLngLiteralMeta,

  /**
   * CREATED: Just created, not sent to backend
   * PERSISTED: Loaded from backend
   * DELETED: PERSISTED and deleted in the frontend
   */
  state: t.union([t.literal('CREATED'), t.literal('PERSISTED'), t.literal('DELETED')])
}, "RouteLocation")

export type RouteLocation = t.TypeOf<typeof RouteLocationMeta>

export const wrapsAroundMidnight = (loc: RouteLocation) => loc.exitTime < loc.arrivalTime;

export const locationsConflict = (loc1: RouteLocation, loc2: RouteLocation) => {
  if (!wrapsAroundMidnight(loc1) && !wrapsAroundMidnight(loc2))
    return loc1.arrivalTime <= loc2.exitTime && loc1.exitTime >= loc2.arrivalTime;
  else if (wrapsAroundMidnight(loc1) && !wrapsAroundMidnight(loc2))
    return loc1.exitTime >= loc2.arrivalTime || loc1.arrivalTime <= loc2.exitTime;
  else if (!wrapsAroundMidnight(loc1) && wrapsAroundMidnight(loc2))
    return loc1.arrivalTime <= loc2.exitTime || loc1.exitTime >= loc2.arrivalTime;
  else
    return true; // If both wrap around, that means they conflict at midnight
}

export const blankRouteLocation = () => ({
  routeLocationId: -1,
  arrivalTime: new Date(),
  exitTime: new Date(),
  stopId: 0,
  coords: {lat: 0, lng: 0},
  state: 'CREATED'
})

export const backendToFrontend = (obj: any, stopId: number) => ({
  ...obj,
  stopId,
  coords: {
    lat: obj.position.latitude,
    lng: obj.position.longitude
  },
  arrivalTime: utcTimeStringToDate(obj.arrivalTime),
  exitTime: utcTimeStringToDate(obj.exitTime),
  state: "PERSISTED"
})

export const frontendToBackend = (pt: RouteLocation, routeId: number) => {
  let result: any = {
    routeId: routeId,
    routeLocationId: pt.routeLocationId > 0 ? pt.routeLocationId : null,
    arrivalTime: pt.arrivalTime,
    exitTime: pt.exitTime,
    lat: pt.coords.lat,
    lng: pt.coords.lng,
  };
  return result;
}
