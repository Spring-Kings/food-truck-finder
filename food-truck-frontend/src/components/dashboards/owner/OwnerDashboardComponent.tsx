import React, {useEffect, useState} from "react";
import Router from "next/router";

import {
  Accordion,
  AccordionDetails,
  AccordionSummary,
  Button,
  Card, CardActionArea, CardActions, CardContent, CardHeader,
  CircularProgress,
  Container, Dialog, DialogContent, Grid,
  GridList,
  GridListTile, List, ListItem,
  Typography,
} from "@material-ui/core";

import TruckListComponent from "../TruckListComponent";
import {UserData} from "../../../redux/user/UserReducer";
import {loadTodaysRoute} from "../../../api/RouteLocation";
import {DEFAULT_ERR_RESP} from "../../../api/DefaultResponses";
import {RouteLocation} from "../../map/route-map/RouteLocation";
import TruckLocationMapComponent from "../../map/truck_location_map/TruckLocationMapComponent";
import {useFlexGrowStyles} from "../../theme/FoodTruckThemeProvider";
import TruckCardComponent from "../../truck/TruckCardComponent";
import {StyledDialogTitle} from "../../util/StyledDialogTitle";
import SendNotificationComponent from "../../notifications/SendNotificationComponent";
import CreateTruckForm from "../../CreateTruckForm";

interface OwnerDashboardProps {
  data: UserData | undefined;
  readonly [x: string]: any;
}

function OwnerDashboardComponent(props: OwnerDashboardProps) {
  const classes = useFlexGrowStyles();
  const [inError, setInError]: [string | null, any] = useState(null);
  const [routePts, setRoutePts]: [RouteLocation[], any] = useState([]);
  const [creatingTruck, setCreatingTruck]: [boolean, any] = useState(false);

  const viewTruck = (id: number): void => {
    Router.replace(`/truck/${id}`);
  };

  const toUserDashboard = (): void => {
    Router.replace("/dashboard/user");
  }

  useEffect(() => {
    props.loadUserFromBackend().then(
      (_response: any) => setInError(null),
      (err: any) => setInError(err)
    );
  }, []);

  useEffect(() => {
    props.data?.ownedTrucks?.forEach(async truck => {
      try {
        const pts = await loadTodaysRoute(truck.id, DEFAULT_ERR_RESP);
        setRoutePts(routePts.concat(pts));
      } catch (err) {
        setInError(err);
      }
    });
  }, [props.data?.ownedTrucks]);

  // Be safe: don't show to people who don't deserve it. Of course you can hack and bypass it,
  // but it looks better to our "customers"
  // Uh, who's paying for this thing again?
  if (props.data === undefined)
    return (
      <Container>
        <CircularProgress />
      </Container>
    );
  else if (inError)
    return (
      <Container>
        <Typography variant="h2">Error</Typography>
        <Typography>{inError}</Typography>
      </Container>
    );
  else if (!props.data.ownedTrucks)
    return (
      <Container>
        <Typography variant="h2">Oops!</Typography>
        <Typography>
          It seems you got to this corner of our site by mistake. Please
          return to the home menu
        </Typography>
        <Button variant="contained" onClick={() => Router.replace("/")}>
          Home
        </Button>
      </Container>
    );

  // Return true UI
  // const oldUI = (
  //   <>
  //     {/** Props IDd using: https://material-ui.com/components/grid/ */}
  //     <GridList
  //       cols={5}
  //       style={{
  //         height: "100vh",
  //         width: "100%",
  //       }}
  //     >
  //       {/** Side list */}
  //       <GridListTile cols={1} style={{ height: "100vh" }}>
  //         {/* Return to user dashboard */}
  //         <Card>
  //           <Button
  //             variant="contained"
  //             color="primary"
  //             onClick={toUserDashboard}
  //           >
  //             User Dashboard
  //           </Button>
  //         </Card>
  //
  //         {/* Subscribe list */}
  //         <Accordion>
  //           <AccordionSummary>My Trucks</AccordionSummary>
  //           <AccordionDetails>
  //             <TruckListComponent
  //               trucks={props.data.ownedTrucks}
  //               tail={
  //                 <Card>
  //                   <Button
  //                     variant="contained"
  //                     onClick={() => Router.replace("/create-truck")}
  //                   >
  //                     New...
  //                   </Button>
  //                 </Card>
  //               }
  //               handleTruckIcon={<Typography>View</Typography>}
  //               handleTruck={viewTruck}
  //             />
  //           </AccordionDetails>
  //         </Accordion>
  //
  //         {/* Open manager page */}
  //         <Card>
  //           <Button
  //             variant="contained"
  //             onClick={() => Router.replace("/manage-trucks")}
  //           >
  //             Manage trucks
  //           </Button>
  //         </Card>
  //       </GridListTile>
  //
  //       {/** Where the map would be */}
  //       <GridListTile cols={4} style={{ height: "100vh" }}>
  //         <TruckLocationMapComponent locations={routePts} />
  //       </GridListTile>
  //     </GridList>
  //   </>
  // );

  return (
    <Grid container>
      <Grid item container direction="row" justify="flex-start" alignItems="flex-start" className={classes.root}>
        <Grid item>
          <Button
            variant="contained"
            onClick={toUserDashboard}
          >
            User Dashboard
          </Button>
        </Grid>
        <Grid item>
          <Button onClick={() => setCreatingTruck(true)}>
            Create Truck
          </Button>
          <Dialog open={creatingTruck}
                  fullWidth
                  maxWidth="md">
            <StyledDialogTitle onClose={() => setCreatingTruck(false)}>
              Create Truck
            </StyledDialogTitle>
            <DialogContent>
              <CreateTruckForm/>
            </DialogContent>
          </Dialog>
        </Grid>
      </Grid>
      <Grid item container direction="row" alignItems="flex-start" spacing={0} className={classes.root}>
        <Grid item direction="column" xs={5}>
          <Container style={{maxHeight: '50vh', overflow: 'auto'}}>
            <List disablePadding>
              {
                props.data?.ownedTrucks?.map(truck => (
                  <ListItem style={{minWidth: '100%'}} disableGutters>
                    <TruckCardComponent id={truck.id} userOwnsTruck={true}/>
                  </ListItem>
                ))
              }
            </List>
          </Container>
        </Grid>
        <Grid item xs={7}>
          <TruckLocationMapComponent locations={routePts} height="50vh"/>
        </Grid>
      </Grid>
    </Grid>
  );
}

export default OwnerDashboardComponent;
