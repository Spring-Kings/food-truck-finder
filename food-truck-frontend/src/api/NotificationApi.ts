import api from "../util/api";
import {parse} from "../util/type-checking";
import {Notification, NotificationMeta} from "../domain/Notification";

export const getNotifications = async (): Promise<Notification[] | null> => {
  let notifications: Notification[] = [];
  const response = await api.get('/notifications');
  if (response.data == null)
    return null;
  if (response.data) {
    for (const n of response.data) {
      const parsed = parse(NotificationMeta, n);
      if (parsed !== null)
        notifications.push(parsed)
    }
  }
  return notifications;
}

export const sendNotification = async (
  truckId: number,
  message: string,
  onFail?: (res: any) => void
) => {
  let data = {
    truckId: truckId,
    message: message,
  };
  await api.post(`/truck/notification`, data)
    .catch(onFail);
}

export const deleteNotification = async (
  notificationId: number,
  onFail?: (res: any) => void
) => {
  await api.delete(`/notification/${notificationId}/delete`, {})
    .catch(onFail);
}

export const setNotificationAsRead = async (
  notificationId: number,
  isRead: boolean,
  onFail?: (res: any) => void
) => {
  await api.put(`/notification/read`, {
    notificationId: notificationId,
    isRead: isRead
  })
    .catch(onFail);
}