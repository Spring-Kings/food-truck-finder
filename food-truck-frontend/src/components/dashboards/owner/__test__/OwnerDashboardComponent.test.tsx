import React from "react";

import * as route from "../../../../api/RouteLocation";
import OwnerDashboardComponent from "../OwnerDashboardComponent";
import user_data from "../../../../util/test_util/user_data.json";
import { shallow } from "enzyme";
import {
  RouteLocation,
  RouteLocationState,
} from "../../../map/route-map/RouteLocation";

const MOCK_REQUEST = jest.fn();

const LOC: RouteLocation = {
  routeLocationId: 0,
  arrivalTime: new Date(),
  exitTime: new Date(),
  stopId: 0,
  coords: { lat: 0, lng: 0 },
  state: RouteLocationState.PERSISTED,
};

describe("Test owner dashboard", () => {
  /* Mock out route API */
  beforeAll(() => {
    jest.mock("../../../../api/RouteLocation", () => {
      loadTodaysRoute: MOCK_REQUEST;
    });
  });

  /* Clear the mock request, to ensure no cross-test contamination */
  afterEach(() => {
    MOCK_REQUEST.mockClear();
  });

  /* Snapshot */
  test("Matches snapshot", () => {
    MOCK_REQUEST.mockImplementationOnce(() => [LOC]);
    const component = shallow(
      <OwnerDashboardComponent
        data={user_data}
        loadUserFromBackend={() => {}}
      />
    );
    expect(component).toMatchSnapshot();
  });
});
