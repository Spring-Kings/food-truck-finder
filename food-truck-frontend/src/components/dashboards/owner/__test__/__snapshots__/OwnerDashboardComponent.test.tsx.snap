// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test owner dashboard Matches snapshot 1`] = `ShallowWrapper {}`;
